domain/src/main/java/org/oppia/domain/topic/TopicController.kt

```diff
@@ -186,6 +220,170 @@ class TopicController @Inject constructor(
     )
   }
 
+  fun getCompletedStoryList(profileId: ProfileId): LiveData<AsyncResult<CompletedStoryList>> {
+    return dataProviders.convertToLiveData(retrieveCompletedStoryListDataProvider(profileId))
+  }
+
+  fun getOngoingTopicList(profileId: ProfileId): LiveData<AsyncResult<OngoingTopicList>> {
+    return dataProviders.convertToLiveData(retrieveOngoingTopicListDataProvider(profileId))
+  }
+
+  private fun retrieveCompletedStoryListDataProvider(profileId: ProfileId): DataProvider<CompletedStoryList> {
+    return dataProviders.transformAsync(
+      TRANSFORMED_GET_COMPLETED_STORIES_PROVIDER_ID,
+      storyProgressController.retrieveCacheStore(profileId)
```

Comment:
Could we do this in a way that doesn't require the cache sore to be exposed? That feels like an implementation detail of StoryProgressController, so TopicController ought to not know about it.

Reply:
If I transfer this to `StoryProgressController` than in that case, I will need to introduced `TopicController` inside `StoryProgressController` and that was giving me following error:

```
error: [Dagger/DependencyCycle] Found a dependency cycle:
public abstract interface ApplicationComponent {
                ^
      org.oppia.domain.topic.TopicController is injected at
          org.oppia.domain.topic.StoryProgressController(…, arg3, …)
      org.oppia.domain.topic.StoryProgressController is injected at
          org.oppia.domain.topic.TopicController(…, arg3)
      org.oppia.domain.topic.TopicController is injected at
          org.oppia.app.topic.lessons.TopicLessonsFragmentPresenter(…, topicController)
      org.oppia.app.topic.lessons.TopicLessonsFragmentPresenter is injected at
          org.oppia.app.topic.lessons.TopicLessonsFragment.topicLessonsFragmentPresenter
      org.oppia.app.topic.lessons.TopicLessonsFragment is injected at
          org.oppia.app.fragment.FragmentComponent.inject(org.oppia.app.topic.lessons.TopicLessonsFragment) [org.oppia.app.application.ApplicationComponent → org.oppia.app.activity.ActivityComponent → org.oppia.app.fragment.FragmentComponent]
```

Reply:
What dependency of TopicController is needed here? If it’s the data provider ID, I think you can move that to the progress controller, too.

Reply:
`retrieveStory ` is needed from Topic controller

Reply:
I think if you change the logic to provide a data provider that gives all story progress objects, you can avoid exposing the database file and transform it here to retrieve the story and filter by last completed exploration ID. 

Also, I suggest adding story and topic IDs as proto fields to their corresponding progress objects if that helps simplify things.

Reply:
@BenHenning Sounds good, then I will update it accordingly

Reply:
Change implementation such that we do not expose cache anymore and introduced topicId and storyId in progress proto to make things simpler.